File: DBSCAN-draft.cpp
#include <iostream>
#include <vector>
#include <cmath>
#include <set>

using namespace std;

// Function to calculate Euclidean distance between two points
double euclideanDistance(const vector<double>& point1, const vector<double>& point2) {
    double sum = 0.0;
    for (size_t i = 0; i < point1.size(); ++i) {
        sum += (point1[i] - point2[i]) * (point1[i] - point2[i]);
    }
    return sqrt(sum);
}

// DBSCAN Algorithm
class DBSCAN {
public:
    DBSCAN(double eps, int minPts) : eps(eps), minPts(minPts), clusterID(0) {}

    // Runs the DBSCAN algorithm
    vector<int> run(const vector<vector<double>>& data) {
        labels.assign(data.size(), -1); // Initially, all points are marked as noise (-1)

        for (size_t i = 0; i < data.size(); ++i) {
            if (labels[i] != -1) continue; // Skip already visited points

            vector<int> neighbors = regionQuery(data, i);
            if (neighbors.size() < minPts) {
                labels[i] = -1; // Mark as noise
            } else {
                clusterID++; // Start a new cluster
                expandCluster(data, i, neighbors);
            }
        }

        return labels; // Return the cluster labels
    }

private:
    double eps;
    int minPts;
    int clusterID;
    vector<int> labels;

    // Expand cluster recursively for a core point
    void expandCluster(const vector<vector<double>>& data, int pointIdx, vector<int>& neighbors) {
        labels[pointIdx] = clusterID;

        size_t i = 0;
        while (i < neighbors.size()) {
            int neighborIdx = neighbors[i];

            if (labels[neighborIdx] == -1) { // If it is marked as noise, include it in the cluster
                labels[neighborIdx] = clusterID;
            }

            if (labels[neighborIdx] == -1 || labels[neighborIdx] == 0) {
                labels[neighborIdx] = clusterID;

                vector<int> newNeighbors = regionQuery(data, neighborIdx);
                if (newNeighbors.size() >= minPts) {
                    neighbors.insert(neighbors.end(), newNeighbors.begin(), newNeighbors.end());
                }
            }

            i++;
        }
    }

    // Find all neighbors within eps distance of a point
    vector<int> regionQuery(const vector<vector<double>>& data, int pointIdx) {
        vector<int> neighbors;
        for (size_t i = 0; i < data.size(); ++i) {
            if (euclideanDistance(data[pointIdx], data[i]) <= eps) {
                neighbors.push_back(i);
            }
        }
        return neighbors;
    }
};

int main() {
    // Example Dataset (2D Points)
    vector<vector<double>> data = {
        {1.0, 2.0}, {2.0, 2.0}, {2.0, 3.0}, {8.0, 7.0},
        {8.0, 8.0}, {25.0, 80.0}, {8.0, 6.0}, {7.0, 7.0}
    };

    // DBSCAN Parameters
    double eps = 2.0; // Radius of neighborhood
    int minPts = 2;   // Minimum points to form a cluster

    // Run DBSCAN
    DBSCAN dbscan(eps, minPts);
    vector<int> labels = dbscan.run(data);

    // Print Cluster Labels
    cout << "Cluster Labels:" << endl;
    for (size_t i = 0; i < labels.size(); ++i) {
        cout << "Point " << i << ": Cluster " << labels[i] << endl;
    }

    return 0;
}

File: DBSCAN-draft.py
import numpy as np

def dbscan(data, eps, minPts):
    """
    Density-Based Spatial Clustering of Applications with Noise (DBSCAN)
    
    Parameters:
        data (numpy.ndarray): 2D array of data points [n_samples, n_features]
        eps (float): Radius of neighborhood
        minPts (int): Minimum number of points to form a core point

    Returns:
        labels (list): Cluster labels for each point (-1 represents noise)
    """
    # Initialize
    labels = [-1] * len(data)  # Initially, all points are labeled as noise (-1)
    cluster_id = 0
    
    def region_query(point_idx):
        """Find all points within eps distance of a given point."""
        neighbors = []
        for idx, other_point in enumerate(data):
            if np.linalg.norm(data[point_idx] - other_point) <= eps:
                neighbors.append(idx)
        return neighbors
    
    def expand_cluster(point_idx, neighbors, cluster_id):
        """Expand the cluster by recursively adding density-reachable points."""
        labels[point_idx] = cluster_id  # Assign current point to the cluster
        
        i = 0
        while i < len(neighbors):
            neighbor_idx = neighbors[i]
            if labels[neighbor_idx] == -1:  # If it's noise, make it part of the cluster
                labels[neighbor_idx] = cluster_id
            elif labels[neighbor_idx] == 0:  # If it hasn't been visited yet
                labels[neighbor_idx] = cluster_id
                new_neighbors = region_query(neighbor_idx)
                if len(new_neighbors) >= minPts:  # Add neighbors if it's a core point
                    neighbors += new_neighbors
            i += 1
    
    for point_idx in range(len(data)):
        if labels[point_idx] != -1:  # Skip visited points
            continue
        neighbors = region_query(point_idx)
        if len(neighbors) < minPts:
            labels[point_idx] = -1  # Mark as noise
        else:
            cluster_id += 1  # Start a new cluster
            expand_cluster(point_idx, neighbors, cluster_id)
    
    return labels

# Example Usage
if __name__ == "__main__":
    # Example 2D dataset
    data = np.array([
        [1, 2], [2, 2], [2, 3], [8, 7],
        [8, 8], [25, 80], [8, 6], [7, 7]
    ])

    # DBSCAN parameters
    eps = 2  # Neighborhood radius
    minPts = 2  # Minimum points to form a cluster

    # Run DBSCAN
    cluster_labels = dbscan(data, eps, minPts)

    # Display results
    print("Cluster labels:", cluster_labels)
